### ISSUE TITLE
How to properly support Elementors WooCommerce product grid when custom wrappers are used by the theme (woocommerce_before_shop_loop_item)

### ISSUE THREAD

**Issue by MapSteps** [elementor/elementor#16057]
## Prerequisites

- [x ] I have searched for similar issues in both open and closed tickets and cannot find a duplicate.
- [x ] The issue still exists against the latest stable version of Elementor.

## Description

Hey friends,

I'm struggeling to keep up supporting the very latest changes in Elementor.

With my Page Builder Framework theme, I am using custom wrappers to display WooCommerce products in a responsive grid like this:

```
/**
 * Construct starting product loop wrapper.
 *
 * @param string $start The loop start output.
 */
function wpbf_woo_product_loop_start( $start ) {

	/**
	 * Only use this if you 100% know what you are doing.
	 * Allows us and 3rd parties to remove Page Builder Frameworks WooCommerce product loop wrappers.
	 */
	if ( ! apply_filters( 'wpbf_woo_product_loop_wrapper', true ) ) {
		return $start;
	}

	// This should help with better supporting Elementor.
	// Since this class is only present on loops created with Elementor, this is save to add globally
	// if Elementor is installed.
	if ( class_exists( '\Elementor\Plugin' ) ) {
		$elementor_support = ' elementor-grid';
	} else {
		$elementor_support = '';
	}

	$products_per_row   = json_decode( get_theme_mod( 'woocommerce_loop_products_per_row' ), true );
	$desktop_breakpoint = wpbf_get_theme_mod_value( $products_per_row, 'desktop', 4, true );
	$tablet_breakpoint  = wpbf_get_theme_mod_value( $products_per_row, 'tablet', 3, true );
	$mobile_breakpoint  = wpbf_get_theme_mod_value( $products_per_row, 'mobile', 1, true );
	$grid_gap           = get_theme_mod( 'woocommerce_loop_grid_gap', 'large' );

	return '<ul class="wpbf-grid wpbf-grid-' . esc_attr( $grid_gap ) . ' wpbf-grid-1-' . esc_attr( $mobile_breakpoint ) . ' wpbf-grid-small-1-' . esc_attr( $tablet_breakpoint ) . ' wpbf-grid-large-1-' . esc_attr( $desktop_breakpoint ) . ' products' . $elementor_support . ' columns-' . esc_attr( $desktop_breakpoint ) . '">';

}
add_filter( 'woocommerce_product_loop_start', 'wpbf_woo_product_loop_start', 0 );
```
I noticed that Elementor also seems to be adding custom wrappers so I brought in the class `elementor-grid` in an attempt to better support Elementor and let the page builder "take over" when it is active.

This has worked fine at first glance, but here's a deeper issue with this.

All this works great if an archive or product page was built with Elementor. It also works great if no theme-building capabilities are used on the archive or product pages.
Though, I just noticed if we have an Elementor header set across the entire site for instance, the `elementor-grid` class receives some styling from `frontend.min.css` and therefore breaks the styling again because the related inline-styles are missing.

Meaning: It is not safe for us to simply add the `elementor-grid` class across the board to our wrapper class.

A possible solution would be if there was a way for us to detect not only if a page was built with Elementor, but if ANY parts of a particular page or archive have parts built with Elementor in it.

If there was a way for us to know that, we can simply use our `wpbf_woo_product_loop_wrapper` (check above) to bypass our wrappers and let Elementor take over.

Though, there seems to be no proper way to detect that. At least I couldn't find one.
Sure, on regular pages built with E, we can check for:

`\Elementor\Plugin::$instance->documents->get( $post->ID )->is_built_with_elementor();`

but there doesn't seem to be a way to detect if a page has only some parts built with Elementor or not. Like a custom Header for instance.

Having a simple body class available to us would do the trick. The body class can be unique and only applied if there is _anything_ built with Elementor present on the page or archive.

This is just my first thought of course, maybe there's an easier/better way to do what I'm trying to achieve.
I'd love your thoughts on this.

Thank you so much for your support.

## Steps to reproduce

1. Install Page Builder Framework Theme
2. With Elementors theme-building capabilities, create a shop page template and make it apply to all shop archives
3. Check the shop page and see how everytyhing looks great
4. Set the shop page template to draft
5. Create a custom header with Elementor and apply it to the entire site
6. visit the shop again and see how the grid is broken

## Isolating the problem

- [x ] This bug happens with only Elementor plugin active (and Elementor Pro).
- [ ] This bug happens with a default WordPress theme active.
- [ x] I can reproduce this bug consistently using the steps above.

## Environment

<details>
<summary>System Info</summary>
```		
== Server Environment ==
	Operating System: Linux
	Software: nginx/1.21.0
	MySQL version: mariadb.org binary distribution v10.5.5
	PHP Version: 7.4.18
	PHP Max Input Vars: 1000
	PHP Max Post Size: 128M
	GD Installed: Yes
	ZIP Installed: Yes
	Write Permissions: All right
	Elementor Library: Connected

== WordPress Environment ==
	Version: 5.8
	Site URL: http://repository.local
	Home URL: http://repository.local
	WP Multisite: No
	Max Upload Size: 128 MB
	Memory limit: 256M
	Permalink Structure: /%postname%/
	Language: en-US
	Timezone: 0
	Debug Mode: Inactive

== Theme ==
	Name: Page Builder Framework - Child
	Version: 1.3
	Author: David Vongries
	Child Theme: Yes
	Parent Theme Name: Page Builder Framework
	Parent Theme Version: 2.6.20
	Parent Theme Author: David Vongries

== User ==
	Role: administrator
	WP Profile lang: en_US
	User Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:91.0) Gecko/20100101 Firefox/91.0

== Active Plugins ==
	Elementor
		Version: 3.4.1
		Author: Elementor.com

	Elementor Pro
		Version: 3.3.7
		Author: Elementor.com

	Page Builder Framework Premium Addon
		Version: 2.7.4
		Author: David Vongries

	Ultimate Dashboard
		Version: 3.5
		Author: David Vongries

	WooCommerce
		Version: 5.6.0
		Author: Automattic


== Elementor - Compatibility Tag ==
	
	Elementor Pro: Compatibility not specified

== Elementor Pro - Compatibility Tag ==
	


== Elements Usage ==
	
	header : 1
		icon-list : 1
		image : 1
		nav-menu : 1
	wp-page : 1


== Log ==
	
Log: showing 12 of 122021-08-16 10:05:47 [info] Elementor data updater process has been completed. [array (
  'plugin' => 'Elementor Pro',
  'from' => '3.3.4',
  'to' => '3.3.7',
)]
2021-08-16 13:38:23 [info] Elementor data updater process has been queued. [array (
  'plugin' => 'Elementor',
  'from' => '3.3.1',
  'to' => '3.4.0',
)]
2021-08-16 13:38:50 [info] elementor::elementor_updater Started 
2021-08-16 13:38:50 [info] Elementor/Upgrades - _on_each_version Start  
2021-08-16 13:38:51 [info] Elementor/Upgrades - _on_each_version Finished 
2021-08-16 13:38:51 [info] Elementor/Upgrades - _on_each_version Start  
2021-08-16 13:38:51 [info] Elementor data updater process has been completed. [array (
  'plugin' => 'Elementor',
  'from' => '3.3.1',
  'to' => '3.4.0',
)]
2021-08-19 06:52:17 [info] Elementor data updater process has been queued. [array (
  'plugin' => 'Elementor',
  'from' => '3.4.0',
  'to' => '3.4.1',
)]
2021-08-19 06:52:26 [info] elementor::elementor_updater Started 
2021-08-19 06:52:26 [info] Elementor/Upgrades - _on_each_version Start  
2021-08-19 06:52:27 [info] Elementor/Upgrades - _on_each_version Finished 
2021-08-19 06:52:27 [info] Elementor data updater process has been completed. [array (
  'plugin' => 'Elementor',
  'from' => '3.4.0',
  'to' => '3.4.1',
)]
```
</details>

**Comment by MapSteps** [comment]
If the body class approach makes sense, it could be called something like `elementor-is-active`. Then, we can check for if it is present and remove our wrappers.

**Comment by udidol** [comment]
Hi @MapSteps,
The reason for this is that in addition to adding the `elementor-grid` class to the products `<ul>` tag in Elementor Pro's WooCommerce widgets, we also add designated CSS and controls along with each widget that makes the grids work.

You should add appropriate CSS to your own code to make sure it works.
I don't know what your CSS looks like, but what you can do is compare the CSS attached to the output for the product loop in an Elementor widget to the output of your own customized product loop.

You can look at the differences from the `.woocommerce` wrapper and downwards to the `ul` and `li` elements, spot the differences and add CSS of your own to bridge those gaps.

**Comment by MapSteps** [comment]
Hi @udidol,

Thanks for pointing me in that direction. That might actually work. I'll report back after testing :)

**Comment by MapSteps** [comment]
@udidol, I think I got it figured out.

In `frontend.min.css`, Elementor sets the styling for `.elementor-grid` to `display: grid`. That's what's breaking the layout for us as we are using flexbox here.

I have now added CSS to the theme to make sure `.elementor-grid` is set to `display: flex` by default.

As you have mentioned, Elementor then extends `.elementor-grid` when specific widgets are used. This will then also override the `display: flex` styling that I have added.

Technically speaking, we are now following the same logic/approach.

Sweet! I'm closing this for now and will keep an eye on wether we get more reports on this after I pushed the update.

Best,
David
